python/ray/data/preprocessors/discretizer.py

```diff
@@ -0,0 +1,360 @@
+
+        >>> import pandas as pd
+        >>> import ray
+        >>> from ray.data.preprocessors import CustomKBinsDiscretizer
+        >>> df = pd.DataFrame({
+        ...     "value_1": [0.2, 1.4, 2.5, 6.2, 9.7, 2.1],
+        ...     "value_2": [10, 15, 13, 12, 23, 25],
+        ... })
+        >>> ds = ray.data.from_pandas(df)  # doctest: +SKIP
+        >>> discretizer = CustomKBinsDiscretizer(
+        ...     columns=["value_1", "value_2"],
+        ...     bins=[0, 1, 4, 10, 25]
+        ... )
+        >>> discretizer.transform(ds).to_pandas()  # doctest: +SKIP
+           value_1  value_2
+        0        0        2
+        1        1        3
+        2        1        3
+        3        2        3
+        4        2        3
+        5        1        3
+
+        You can also specify different bin edges per column.
+
+        >>> discretizer = CustomKBinsDiscretizer(
+        ...     columns=["value_1", "value_2"],
+        ...     bins={"value_1": [0, 1, 4], "value_2": [0, 18, 35, 70]},
+        ... )
+        >>> discretizer.transform(ds).to_pandas()  # doctest: +SKIP
+           value_1  value_2
+        0        0        0
+        1        1        0
+        2        1        0
+        3        1        0
+        4        1        1
+        5        1        1
+
+
+    Args:
+        columns: The columns to discretize.
+        bins: Defines custom bin edges. Can be either an interable of numbers,
+            a ``pd.IntervalIndex``, or a dict mapping columns to either of them.
+            Note that ``pd.IntervalIndex`` for bins must be non-overlapping.
+        right: Indicates whether bins includes the rightmost edge or not.
+        include_lowest: Whether the first interval should be left-inclusive
+            or not.
+        duplicates: Can be either 'raise' or 'drop'. If bin edges are not unique,
+            raise ``ValueError`` or drop non-uniques.
+        dtypes: An optional dictionary that maps columns to ``pd.CategoricalDtype``
+            objects or ``np.integer`` types. If you don't include a column in ``dtypes``
+            or specify it as an integer dtype, the outputted column will consist of
+            ordered integers corresponding to bins. If you use a
+            ``pd.CategoricalDtype``, the outputted column will be a
+            ``pd.CategoricalDtype`` with the categories being mapped to bins.
+            You can use ``pd.CategoricalDtype(categories, ordered=True)`` to
+            preserve information about bin order.
+
+    .. seealso::
+
+        :class:`UniformKBinsDiscretizer`
+            If you want to bin data into uniform width bins.
+    """
+
+    def __init__(
+        self,
+        columns: List[str],
+        bins: Union[
+            Iterable[float],
+            pd.IntervalIndex,
+            Dict[str, Union[Iterable[float], pd.IntervalIndex]],
+        ],
+        *,
+        right: bool = True,
+        include_lowest: bool = False,
+        duplicates: str = "raise",
```

Comment:
```suggestion
        duplicates: Literal["raise", "drop"] = "raise",
```

Reply:
Literals are not supported in python<3.8, which we still target